Replace CICE start date calculations

blimlim opened the pull request:
This PR covers the first part of #466. It tries to simplify the setting of CICE's start date in ACCESS-ESM1.5 simulations, in order to avoid inconsistency bugs and to make the restart timing more transparent to the user. 

The previous calculations for setting the cice start date used:
``` 
control_dir/ice/input_ice.nml
--------------------------------
&coupling
init_date=00010101 (control_init_date)
...
```
and 
``` 
restart_dir/ice/input_ice.nml
--------------------------------
&coupling
runtime0=3155673600 (res_runtime0)
runtime=0 (res_runtime)
/
``` 

in order to set the start date of a new simulation, payu would calculate a start date as `inidate = control_init_date + res_runtime0 + res_runtime` (with required unit conversions).


It would then populate the work directory copy of `input_ice.nml` with values for the model to use for the new simulation:
``` 
work_dir/ice/input_ice.nml
--------------------------------
&coupling
inidate = control_init_date + res_runtime0 + res_runtime
init_date = control_init_date
runtime0= res_runtime0 + res_runtime
runtime= run duration calculated from config.yaml
/
```

---
This PR replaces  `restart_dir/ice/input_ice.nml` with a new file `restart_dir/ice/restart_date.nml` containing:
```
restart_dir/ice/restart_date.nml
---------------------------------
&coupling
init_date=10101 (res_init_date)
inidate=1010101 (res_inidate)
/
```
During setup, payu then fills the work directory copy of `input_ice.nml` with:

``` 
work_dir/ice/input_ice.nml
--------------------------------
&coupling
inidate = res_inidate
init_date = res_init_date
runtime0 = res_inidate - res_init_date (in seconds)
runtime = run duration calculated from config.yaml
/
```

My goal here was to ensure that cice receives the exact same values of `init_date`, `inidate`, `runtime`, `runtime0` with the new calculations that it would have received with the old calculations, which is why the updates use the `init_date` and calculate the number of seconds between it and the `inidate`. It would be simpler to just set both `init_date` and `inidate` to the new start date, and set `runtime0=0`, however this would give different information to `cice`.

For some reason I thought it would be good to read the `init_date` from the new restart file (I forget why...), though this could be moved back to reading it from the control directory if everyone thinks it is cleaner. 

Backwards compatibility will be broken with this PR, and if merged, we will have to add `restart_date.nml` files to the `vk83` restart directories.

Future work would making a separate function for reading the start date, which could then be used for consistency checks between the different component's start dates, though I'll defer that to a future PR/issue. 

---
For the review, any suggestions on structure, clarity, and also correctness of the calculations would be very welcome!

pep8speaks commented:
Hello @blimlim! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`payu/calendar.py`](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py):

> [Line 94:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L94): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 104:30](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L104): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 113:1](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L113): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 114:5](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L114): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 122:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L122): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 125:1](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L125): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 129:73](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L129): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 130:64](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L130): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 142:13](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L142): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 142:15](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L142): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 143:14](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L143): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 143:16](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L143): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 144:12](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L144): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 144:14](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L144): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 145:13](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L145): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 145:15](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L145): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 146:15](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L146): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 146:17](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L146): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 147:15](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L147): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 147:17](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L147): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 148:17](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L148): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 148:19](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/calendar.py#L148): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

* In the file [`payu/models/access.py`](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py):

> [Line 61:1](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L61): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 104:78](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L104): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 121:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L121): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 141:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L141): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 142:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 149:72](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L149): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 159:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L159): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 178:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L178): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 179:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L179): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 185:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L185): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 191:17](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L191): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 236:1](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L236): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 241:73](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L241): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 267:65](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L267): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 269:65](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L269): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 269:67](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/payu/models/access.py#L269): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

* In the file [`test/test_calendar.py`](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py):

> [Line 5:80](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 157:1](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L157): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 161:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 161:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 161:50](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 161:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 161:59](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 161:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L161): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:50](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:59](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 162:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L162): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:50](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:59](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 167:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L167): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:50](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:59](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 168:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L168): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 170:20](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L170): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 170:23](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L170): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 173:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 173:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 173:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 173:54](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 173:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 173:63](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L173): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:54](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 174:63](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L174): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:54](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 179:63](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L179): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:54](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 180:63](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L180): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:50](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:52](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:59](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:61](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:39](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:41](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:53](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:55](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:62](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 186:64](https://github.com/payu-org/payu/blob/c1a0ef500068ba7d47e3877f3ed5aaa5d0ff2e8a/test/test_calendar.py#L186): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

blimlim commented:
Oh my bad. I'll fix these PEP issues and ping everyone when done

blimlim commented on payu/calendar.py:
```diff
@@ -90,6 +90,68 @@ def calculate_leapdays(init_date, final_date):
+
+
+def date_to_cftime(date, calendar):
+    """
```
Let me know if you are aware of any built in ways to do this – I couldn't find any from a quick search.

aidanheerdegen replied:
Judging from this `xarray` function you an just subtract one `CFTime` object from another 

https://github.com/pydata/xarray/blob/main/xarray/core/resample_cftime.py#L443-L480

And yeah the code backs that up

https://github.com/Unidata/cftime/blob/master/src/cftime/_cftime.pyx#L1063-L1064

and subtraction just returns a vanilla `timedelta`

https://github.com/Unidata/cftime/blob/master/src/cftime/_cftime.pyx#L1596

blimlim replied:
Very glad that `cftime` does this. I'm making use of this in the `seconds_between_dates` function:
```
  delta = (date_to_cftime(end_date, calendar_str)
             - date_to_cftime(start_date, calendar_str))
```
however am manually converting from `datetime.date` to `cftime.date` to do this calculation. This is because the `start_date` and `end_date`  need to be `datetime.date` objects for other calculations such as `runtime_from_date`.

blimlim commented on payu/models/access.py:
```diff
@@ -52,13 +52,24 @@ def __init__(self, expt, name, config):
 
-        cpl_keys = {'cice': ('input_ice.nml', 'coupling', 'runtime0'),
-                    'matm': ('input_atm.nml', 'coupling', 'truntime0')}
-
```
I moved these from here to the initialisation section, making them model attributes so that they would be accessible when writing the new restart date during `archive`.

blimlim commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
             if model.model_type in ('cice', 'matm'):
 
```
The previous calculations were identical for both `cice` and `matm`. It was simplest to make the same changes for both rather than keep all the old code for `matm` and add the new code separately for `cice`. This means that anyone who wants to run with `matm` would need to add the new restart file. Would this be ok to impose?

aidanheerdegen replied:
I'm of the opinion we should drop support for `matm`.

I have made an issue to this effect

https://github.com/payu-org/payu/issues/486

But for the time being, I wouldn't worry about breaking `matm`. Or we can removed it and rebase this PR and it goes away.

blimlim reviewed (commented):
Just added comments with background on a couple of the choices I made.

aidanheerdegen commented on payu/calendar.py:
```diff
@@ -90,6 +90,68 @@ def calculate_leapdays(init_date, final_date):
+    # TODO: Is it confusing that GREGORIAN means proleptic gregorian?
+    if caltype_int == GREGORIAN:
+        calendar_str = "proleptic_gregorian"
+    elif caltype_int == NOLEAP:
+        calendar_str = "noleap"
```
If this is just a case change, then I'd just do a blanket `.lowercase()` at the beginning to ensure any calendar is correct.

blimlim replied:
This one's actually converting from the [integer constants]( https://github.com/ACCESS-NRI/payu/blob/8f18f245787cafa940a9c397b5173479a40b6273/payu/calendar.py#L7) used in the cice namelists to the calendar strings used by `cftime`.

aidanheerdegen replied:
Ah duh! (Me). Sorry.
```suggestion
    elif caltype_int == NOLEAP:
        calendar_str = "noleap"
```
```suggestion
    elif caltype_int == NOLEAP:
        calendar_str = "noleap"
```
```suggestion
    elif caltype_int == NOLEAP:
        calendar_str = "noleap"
```
```suggestion
    elif caltype_int == NOLEAP:
        calendar_str = "noleap"
```

aidanheerdegen replied:
Ah duh! (me). Sorry.

aidanheerdegen commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
+                            model.cpl_group]
+                    except FileNotFoundError:
+                        print(
+                            "Missing restart date file for model "
+                            f"{model.model_type}",
+                            file=sys.stderr
+                        )
```
Will this report the path that is supposed to exist? I think it is always good to tell the user what the expected behaviour was, which in this case that there be a file of a certain name in a particular location.

anton-seaice replied:
In the text for the FileNoteFoundError if will give the expected path

aidanheerdegen commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
+                            "Values provided: \n"
+                            f"inidate = {start_date_nml['inidate']}\n"
+                            f"init_date = {start_date_nml['init_date']}"
+                        )
```
Yeah this is great info to feedback to the user.

aidanheerdegen commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
+                    )
 
                 else:
-                    total_runtime = 0
```
Good idea, much clearer.

aidanheerdegen commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
                 # Which calendar are we using, noleap or Gregorian.
-                caltype = cpl_nml[cpl_group]['caltype']
-                init_date = cal.int_to_date(cpl_nml[cpl_group]['init_date'])
+                caltype = cpl_nml[model.cpl_group]['caltype']
```
```suggestion
                # Use variable to point directly to coupling namelist group
                cpl_group = cpl_nml[model.cpl_group]
                caltype = cpl_group['caltype']
```
If you want to can make a variable to point directly to the coupling namelist group, which would clean up the later sections which also set variables here. It's a pattern that used elsewhere in the code.

Just a suggestion, up to you if you think it improves or degrades readability.

blimlim replied:
Great idea! I think it makes it easier to read, and added this change in to the latest commit.

aidanheerdegen commented on test/test_calendar.py:
```diff
@@ -150,3 +154,42 @@ def test_parse_date_offset_no_offset_magnitude():
+                (1000 * 365 + 242) * SEC_PER_DAY
+            ),
+        ]
+)
```
Great to add tests, thanks. Just a note that all of the tests use `month=1` and `day=1`, so probably worth changing a couple to something else.

blimlim replied:
Good idea. I've added in tests using different months and days, and also added a test running between year 1 and 9999, which are the minimum and maximum years for `datetime.date`s.

anton-seaice replied:
Thanks - new test looks good :)

aidanheerdegen commented on payu/calendar.py:
```diff
@@ -90,6 +90,68 @@ def calculate_leapdays(init_date, final_date):
+    """
+    # Get the cftime string corresponding to the caltype integer
+
+    # TODO: Is it confusing that GREGORIAN means proleptic gregorian?
```
Well it's confusing that we use `gregorian` when we mean `proleptic_gregorian` and so have to write little work-arounds like this.

aidanheerdegen reviewed (commented):
I'd really really really recommend adding some model driver tests, even just a minimal set that covers the expected behaviour of this code.

There are some examples you can base tests on here

https://github.com/payu-org/payu/blob/master/test/models/

I'd guess the one @jo-basevi wrote would be a good start

https://github.com/payu-org/payu/blob/master/test/models/test_mom_mixin.py

anton-seaice commented on payu/models/access.py:
```diff
@@ -89,61 +100,68 @@ def setup(self):
+                    # Start date of new run
+                    run_start_date = cal.int_to_date(start_date_nml["inidate"])
+
+                    # run_start_date must be after initialisation date
```
Did you test this? Its possible this would just work ?

blimlim replied:
I tried running with 
```    
inidate = 1010101
init_date = 10101
```
and then during `payu setup` forced `runtime0=0`, to test whether it is sufficient to set just the `inidate`.

Some strange things end up happening, e.g. in `iceout085`:
```
(calendar)  idate =      1010101
(calendar)  idate =      2010101
(calendar)  idate =      2010101
```
the year jumped by 100 years, which also impacted the output file names:
```
iceh.0201-01.nc
```
Values like `runtime0` are [showing up](https://github.com/ACCESS-NRI/cice4/blob/e7549ebd2044690a432cc67c1317c81cb194b750/source/ice_restart.F90#L330) in the cice code, and so figured that a safe approach for the new calculations would be to ensure that cice is still provided with the same values of `inidate`, `init_date`, and `runtime0` as before.

anton-seaice replied:
Thanks :)

anton-seaice reviewed (commented):
[input_ice](https://github.com/ACCESS-NRI/access-esm1.5-configs/blame/dev-historical%2Bconcentrations/ice/input_ice.nml) has lots of fields unrelated to restarts in it, so i'm not sure why we are adding another file? Whats the purpose in keeping the timing information on the completed run seperately?

I assume we are changing how runtime0 is calculated
from
`runtime0= res_runtime0 + res_runtime`
to
`runtime0 = res_inidate - res_init_date (in seconds)`
so that the user can change `res_inidate` if they desire (this is the date the current run starts at)?.

> For some reason I thought it would be good to read the init_date from the new restart file (I forget why...), though this could be moved back to reading it from the control directory if everyone thinks it is cleaner.

This is something to do with the warm-start script ... if the config comes from one location and the restart comes from a different location, I don't know which one is sensible to get the restart date in? I guess it depends on the warm start script and also which location its set in other models ?? I'm not really sure why this isn't the only change needed actually ?

blimlim commented:
> [input_ice](https://github.com/ACCESS-NRI/access-esm1.5-configs/blame/dev-historical%2Bconcentrations/ice/input_ice.nml) has lots of fields unrelated to restarts in it, so i'm not sure why we are adding another file? Whats the purpose in keeping the timing information on the completed run seperately?

Previously payu had been copying the whole `input_ice.nml`  into the restart directory, but only the `runtime0` and `runtime` fields from this copy were required. Meanwhile the `inidate` and `init_date` were being ignored, along with all the other timing related fields. The main idea behind the separate `restart_date.nml` file is to try and simplify this, so that the new file holds all the information needed for the run timing, and none of the unnecessary extra fields. Let me know if this approach sounds ok!

Related to this – with the changes, the `input_ice.nml` was not meant to be copied to the restart directory anymore as it isn't needed. I'd forgotten to implement this change, but have added it in to the latest commit.

> to
> `runtime0 = res_inidate - res_init_date (in seconds)`
> so that the user can change `res_inidate` if they desire (this is the date the current run starts at)?.

Unfortunately the restart date is also held by the binary `iced...` restart files, and it needs to be consistent with the `inidate`. This means that changing `res_inidate` won't be enough by itself to change to a completely new restart date. It will make it easier though to create a simulation from a CSIRO restart which currently requires the `runtime0+runtime` calculation to be performed if done manually. Another reason for the change is to try and make it clearer which date cice will be starting at (e.g. so people don't need to calculate `control_init_date + res_runtime0 + res_runtime').

> This is something to do with the warm-start script
That's a good point. If these changes go through, the warm start scripts would need to be updated down the line for them to work with the changed calculations.

> I'm not really sure why this isn't the only change needed actually ?

I think it would work to read the `init_date` from the restart directory instead of the control directory, and have this as the only change, though it might be  a bit less transparent which date the model will restart at. I.e. this would require `runtime_0 + runtime + init_date` calculation. Happy to hear from everyone if either approach is preferrable!

blimlim commented:
> I'd really really really recommend adding some model driver tests, even just a minimal set that covers the expected behaviour of this code.

Good idea. I've had a go at adding some tests, first of the [individual calendar functions](https://github.com/ACCESS-NRI/payu/blob/8f18f245787cafa940a9c397b5173479a40b6273/test/test_calendar.py#L219-L261) used in the calculations, and then a bigger test of the actual driver:
https://github.com/ACCESS-NRI/payu/blob/8f18f245787cafa940a9c397b5173479a40b6273/test/models/test_access.py#L178-L242

The driver test cycles over the `access.setup()` and `access.archive()` steps (currently 1000 times – perhaps excessive) with a year long runtime, and checks that the final end date matches what's expected. It's admittedly a bit slow (~22s for the 1000 cycles), but would be keen to hear if this sounds like a reasonable type of test to ensure that the calculations are working correctly across the different tricky years that can come up?

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=400, month=1, day=1),
+                datetime.datetime(year=401, month=1, day=1),
+                GREGORIAN,
+                366 * SEC_PER_DAY
```
```suggestion
                datetime.datetime(year=400, month=1, day=1),
                datetime.datetime(year=401, month=12, day=31),
                GREGORIAN,
                365 * SEC_PER_DAY
```
This tests that month and day increments are working also

blimlim replied:
Thanks great idea - added this one in!

anton-seaice commented on payu/models/access.py:
```diff
@@ -52,13 +52,24 @@ def __init__(self, expt, name, config):
+                model.cpl_fname = 'input_ice.nml'
+                model.cpl_group = 'coupling'
+                model.runtime0_key = 'runtime0'
+                model.start_date_nml_name = "restart_date.nml"
```
Should keys for "init_date", "inidate" and "runtime" be here rather than hardcoded at line 182 ?

Have we documented what each field means somewhere?

blimlim replied:
That's a good idea – makes it more consistent. I've added in these extra keys, and included comments briefly specifying what their meaning:

https://github.com/ACCESS-NRI/payu/blob/c19133c83109e768236486b0879d09e6f05c3be2/payu/models/access.py#L60-L67

anton-seaice replied:
Nice

anton-seaice commented on payu/models/access.py:
```diff
@@ -208,13 +233,53 @@ def archive(self):
+                work_cpl_grp = work_cpl_nml[model.cpl_group]
+
+                # Timing information on the completed run.
+                run_init_date_int = work_cpl_grp["init_date"]
```
Is it clearer to call this `exp_init_date...` ? Because its start date for the whole experiment ?

blimlim replied:
Yep that's much clearer! I've added this in.

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+        "date_int, expected",
+        [
+            (10101, datetime.date(1, 1, 1)),
+            (100321, datetime.date(10, 3, 21)),
+            (38211130, datetime.date(3821, 11, 30))
```
```suggestion
            (99991231, datetime.date(9999, 12, 31))
```

blimlim replied:
Good idea, added this in

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+    """
+    converted_date = int_to_date(date_int)
+    assert converted_date == expected
+
```
For completeness, this should also test the negative cases:

e.g. what if the year is 0 or year 10000, or the date requested doesn't exist (i.e. 999999)

blimlim replied:
Good question, I've added a separate test to check that the function fails on invalid dates:

https://github.com/ACCESS-NRI/payu/blob/c19133c83109e768236486b0879d09e6f05c3be2/test/test_calendar.py#L229-L239

anton-seaice replied:
Nice

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+                                seconds,
+                                caltype)
+
+    assert runtime == expected
```
Again - we should test the negative cases ... year -1, month 13, etc

blimlim replied:
This is a good point – It turns out that this function currently doesn't guard against bad runtimes being specified in the `config.yaml`. If it's given a negative input it outputs a negative number of seconds for the runtime.

Since this is something that touches a couple other models, do you think it would be ok to defer adding guard rails to a separate issue?

anton-seaice replied:
Ok i raised https://github.com/payu-org/payu/issues/492

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+            (datetime.date(1, 1, 1), 9998, 0, 0, 0,
+             GREGORIAN, (9998 * 365 + 2424) * SEC_PER_DAY)
+        ]
+)
```
I think this would be clearer if you just have the limits in all permutations:

For start_date: the minimum (e.g. datetime.date(1,1,1)) and the maximum (datetime.date(9999,12,31)) 
For run time: the minimum (1 second) and the maximums, (9999,0,0,0) and (9998,12,31,86400)
For calendar: GREGORIAN and NOLEAP

Its about 12 cases.

blimlim replied:
Good idea to test around the edges here. I think it's worth retaining the cases like `(datetime.date(101, 1, 1), 1, 0, 0, 0, GREGORIAN, 365*SEC_PER_DAY)`, as these match what people will generally use for ESM1.5, e.g. 1 year runtimes, starting at the beginning of the year. 

In terms of the max cases, start dates like `datetime.date(9999,12,31)` cause issues, as the `runtime_from_date` function internally creates an end date of `start date + runtime`, which then goes outside the limits of `datetime.date`. I've added in a few different limiting cases though – let me know if you think they cover what we need.

anton-seaice replied:
Im just going to go ahead and assume no paleo folk do a climate run longer than 9999 years !

anton-seaice replied:
This is much clearer, thankyou

anton-seaice replied:
Turns out there was already an issue for the 9999 years thing ! https://github.com/payu-org/payu/issues/305

anton-seaice commented on payu/models/access.py:
```diff
@@ -157,20 +176,26 @@ def setup(self):
+                # work directory namelist.
+                cpl_group["init_date"] = cal.date_to_int(init_date)
+                cpl_group['inidate'] = cal.date_to_int(run_start_date)
+                cpl_group[model.runtime0_key] = previous_runtime
```
When I test this, my `input_ice.nml` gave `    runtime0 = 3155673600.0`

The extra `.0` probably doesn't matter , but is it easy to tidy up?

blimlim replied:
Good catch, I've adjusted the `seconds_between_dates` function to output integers:
```
return int(delta.total_seconds())
```
which should be safe as it's inputs are `datetime.date` objects, and so there shouldn't be any fractions of seconds appearing anywhere.

anton-seaice replied:
nice

anton-seaice commented:
I ran this and it works with the dates look correct for the first couple of months of picontrol. I didn't test trying to change the start date or what happens with a leap year.

anton-seaice commented:
The tests run and pass as-is, so if you want to defer the suggested changes to the tests to after release, that would be ok

blimlim commented:
Thanks for the reviews! Lots of good ideas which help make the changes clearer and also improve the tests. I've added in the different suggestions, and also added an additional test of the `access.py` driver.

https://github.com/ACCESS-NRI/payu/blob/c19133c83109e768236486b0879d09e6f05c3be2/test/models/test_access.py#L261-L269

The previous test which cycled over the `setup` and `archive` stages would not catch cases where calculations in both steps were simultaneously wrong (e.g. if they somehow both used the wrong calendar). The new test focuses on the setup date, checking that the correct runtime is written to the namelists when starting at a range of leap/non-leap years. I think there's some redundancy with this and the more focused tests of `calendar.py` but think it's still helpful to have to make sure that everything comes together correctly in the driver.

anton-seaice commented on test/models/test_access.py:
```diff
@@ -0,0 +1,340 @@
+         (1040101, NOLEAP, 365*SEC_PER_DAY),
+         (3000101, GREGORIAN, 365*SEC_PER_DAY),
+         (3000101, NOLEAP, 365*SEC_PER_DAY),
+         (4000101, GREGORIAN, 366*SEC_PER_DAY),
```
```suggestion
         (4000101, GREGORIAN, 366*SEC_PER_DAY), #not a leap year!
```

anton-seaice replied:
hangon this is a leap year ?

blimlim replied:
Yeah this one should be a leap year!

anton-seaice commented on test/models/test_access.py:
```diff
@@ -0,0 +1,340 @@
+):
+    """
+    The large setup/archive cycling test won't pick up situations 
+    where the calculations dyring setup and archive are simultaneously
```
```suggestion
    where the calculations during setup and archive are simultaneously
```

blimlim replied:
Ah thanks for spotting that!

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,137 @@ def test_parse_date_offset_no_offset_magnitude():
+            # Leap year due to 400 year rule
+            (datetime.date(400, 1, 1), 1, 0, 0, 0, GREGORIAN, 366*SEC_PER_DAY),
+            (datetime.date(500, 1, 1), 1, 0, 0, 0, NOLEAP, 365*SEC_PER_DAY),
+            # Febraury in leap years
```
```suggestion
            # February in leap years
```

blimlim replied:
Good catch!

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,137 @@ def test_parse_date_offset_no_offset_magnitude():
+            # Misc
+            (datetime.date(153, 8, 21), 2, 7, 510, 5321,
+             GREGORIAN, 1453*SEC_PER_DAY + 5321),
+            # Extreme cases, unlikely to ever happen
```
```suggestion
            # Max & min limits
```

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,137 @@ def test_parse_date_offset_no_offset_magnitude():
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, GREGORIAN, 29*SEC_PER_DAY),
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, NOLEAP, 28*SEC_PER_DAY),
+            # Misc
+            (datetime.date(153, 8, 21), 2, 7, 510, 5321,
+             GREGORIAN, 1453*SEC_PER_DAY + 5321),
```
```suggestion
            (datetime.date(1, 1, 1), 0, 0, 0, 86400,
             GREGORIAN, 86400),
```

anton-seaice commented on test/test_calendar.py:
```diff
@@ -150,3 +156,137 @@ def test_parse_date_offset_no_offset_magnitude():
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=12, month=7, day=22),
+                datetime.datetime(year=23, month=3, day=15),
+                GREGORIAN,
+                (10 * 365 + 238) * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=1, month=1, day=1),
+                datetime.datetime(year=1001, month=1, day=1),
+                GREGORIAN,
+                (1000 * 365 + 242) * SEC_PER_DAY
+            ),
```
These two cases are covered sufficiently by the other tests (unless I missed something)
```suggestion
```

blimlim replied:
I've kept in the first test which was based off Aidan's suggestion to test using some mid-month values, and removed the second one – I agree I think that one's covered by the other cases!

anton-seaice commented:
This looks good @blimlim - I just had some tidyups on the testing in the suggestions :)

blimlim commented:
Awesome thanks for those suggestions – very helpful for making the code clearer and improving the tests! I've added those in and let me know if there are any other suggestions.

anton-seaice reviewed (approved):
Good work Spencer !

blimlim commented:
@aidanheerdegen are you happy for this one to be merged, or did you have any other suggestions?
